pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -627,6 +633,315 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+		}
+
+		glog.Infof(
+			"UnmapVolume succeeded for volume %q (OuterVolumeSpecName: %q) pod %q (UID: %q). InnerVolumeSpecName %q. PluginName %q, VolumeGidValue %q",
+			volumeToUnmount.VolumeName,
+			volumeToUnmount.OuterVolumeSpecName,
+			volumeToUnmount.PodName,
+			volumeToUnmount.PodUID,
+			volumeToUnmount.InnerVolumeSpecName,
+			volumeToUnmount.PluginName,
+			volumeToUnmount.VolumeGidValue)
+
+		// Update actual state of world
+		markVolUnmountedErr := actualStateOfWorld.MarkVolumeAsUnmounted(
+			volumeToUnmount.PodName, volumeToUnmount.VolumeName)
+		if markVolUnmountedErr != nil {
+			// On failure, just log and exit
+			glog.Errorf(volumeToUnmount.GenerateErrorDetailed("UnmapVolume.MarkVolumeAsUnmounted failed", markVolUnmountedErr).Error())
+		}
+
+		return nil
+	}, blockVolumePlugin.GetPluginName(), nil
+}
+
+func (og *operationGenerator) GenerateUnmapDeviceFunc(
+	deviceToDetach AttachedVolume,
+	actualStateOfWorld ActualStateOfWorldMounterUpdater,
+	mounter mount.Interface) (func() error, string, error) {
+
+	// Get block volume mapper plugin
+	var blockVolumeMapper volume.BlockVolumeMapper
+	blockVolumePlugin, err :=
+		og.volumePluginMgr.FindMapperPluginBySpec(deviceToDetach.VolumeSpec)
+	if err != nil {
+		return nil, "", deviceToDetach.GenerateErrorDetailed("UnmapDevice.FindMapperPluginBySpec failed", err)
+	}
+	if blockVolumePlugin == nil {
+		return nil, "", deviceToDetach.GenerateErrorDetailed("UnmapDevice.FindMapperPluginBySpec failed to find BlockVolumeMapper plugin. Volume plugin is nil.", nil)
+	}
+	blockVolumeMapper, newMapperErr := blockVolumePlugin.NewBlockVolumeMapper(
+		deviceToDetach.VolumeSpec,
+		nil, /* Pod */
+		volume.VolumeOptions{})
+	if newMapperErr != nil {
+		return nil, "", deviceToDetach.GenerateErrorDetailed("UnmapDevice.NewBlockVolumeMapper initialization failed", newMapperErr)
+	}
+
+	blockVolumeUnmapper, newUnmapperErr := blockVolumePlugin.NewBlockVolumeUnmapper(
+		string(deviceToDetach.VolumeName),
+		"" /* podUID */)
+	if newUnmapperErr != nil {
+		return nil, blockVolumePlugin.GetPluginName(), deviceToDetach.GenerateErrorDetailed("UnmapDevice.NewUnmapper failed", newUnmapperErr)
+	}
+
+	return func() error {
+		globalMapPath, err :=
+			blockVolumeMapper.GetGlobalMapPath(deviceToDetach.VolumeSpec)
+		if err != nil {
+			// On failure, return error. Caller will log and retry.
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice.GetGlobalMapPath failed", err)
+		}
+		refs, err := og.blkUtil.GetDeviceSymlinkRefs(deviceToDetach.DevicePath, globalMapPath)
+		if err != nil {
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice.GetDeviceSymlinkRefs check failed", err)
+		}
+		if len(refs) > 0 {
+			err = fmt.Errorf("The device %q is still referenced from other Pods %v", globalMapPath, refs)
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice failed", err)
+		}
+
+		// The block volume is not referenced from Pods. Release file descriptor lock.
```

Comment:
Can the loopback device cleanup be done at the very end?  Right now, if we remove the loopback device first and subsequent steps fail, then the race condition can be exposed.

Also, what about the scenario where you have multiple maps and unmaps happening on the same volume at the same time?  Could a pod map a volume right before another thread removes the loopback device?  Maybe the loopback remove should not happen until after we mark the volume as unmounted.

Reply:
Actually maybe this is ok.  Does the operation generator ensure that only one operation per volume can be run at a time?

Reply:
@msau42 
I thought reconcier() is running on single thread.
Also during reconcier() loop, it seems that reconcier() ensure these set of operation generator's operations like (1) or (2) are executed as a series. As a result during (1), (2) can't interrupt(doesn't happen) or vice versa.

(1). pod create with vol
- make volume as attached
- make volume as mounted

(2). pod delete with vol
- make volume as unmounted
- make volume as detached (if other pods don't use this volume anymore)

Reply:
Even if only one operation executes at a time, it is possible that it only partially completed, if it encountered an error.

Reply:
@msau42 

>> it is possible that it only partially completed, if it encountered an error.

I agree this possibly happens. Could you let me know your supposed race scenario you mentioned above? I'd like to know what race can be prevented if we move loopback device cleanup to end of the method.

Here are rough steps of UnmapDevice steps.
```
GenerateUnmapDeviceFunc()

(1) check symlink references
  refs>0 -> return err, restart from step.1
  refs =0 -> go step.2

(2) check loopdevice existense
  yes -> remove loopdevice
  no -> warning message

(3) isDeviceOpened check
  still opened -> return err, restart from step.1
  not opened -> go step.4

(4) TearDownDevice
  success -> go step.5
  fail -> return err, restart from step.1

(5) RemoveMapPath(remove globalMapPath dir)
  success or notExist -> go step.6
  fail -> return err, restart from step.1

(6) MarkDeviceAsUnmounted
  success -> return nil(Complete GenerateUnmapDeviceFunc)
  fail -> return err, restart from step.1
```

Reply:
I'm mainly concerned if you remove the loopback device before removing the global mount, would that open up the possibility of the device from being swapped out?

Reply:
Looks like you haven't updated the code to match the operation ordering mentioned above? Could you also add some comments about the reason of this ordering?